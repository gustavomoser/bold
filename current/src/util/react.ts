import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
