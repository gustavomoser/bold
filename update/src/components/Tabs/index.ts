export { TabItem } from './TabItem'
export { Tabs } from './Tabs'

export type { TabItemProps } from './TabItem'
export type { TabsProps } from './Tabs'
