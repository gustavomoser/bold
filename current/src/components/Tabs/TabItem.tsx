import React, { CSSProperties, HTMLAttributes } from 'react'
import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util'

export interface TabItemProps extends Omit<HTMLAttributes<HTMLElement>, 'style'> {
  component?: React.ElementType
  active?: boolean
  disabled?: boolean
  children?: React.ReactNode
  style?: ExternalStyles
  [key: string]: any
}

export function TabItem(props: TabItemProps) {
  const { active, disabled, component, children, style, ...rest } = props
  const { classes, css } = useStyles(createStyles)

  const className =
    css(classes.item, active && classes.active, disabled && classes.disabled) + `${disabled ? ' disabled' : ''}`

  const Cmp = component

  return (
    <li className={css(classes.li, style)}>
      <Cmp role='tab' className={className} {...rest}>
        {children}
      </Cmp>
    </li>
  )
}

TabItem.defaultProps = {
  component: 'span',
}

const createStyles = (theme: Theme) => ({
  li: {
    display: 'inline-block',

    '&:not(:first-child)': {
      marginLeft: '0.25rem',
    },
  } as CSSProperties,
  item: {
    display: 'inline-block',
    textDecoration: 'none',
    color: theme.pallete.text.main,
    fontWeight: 'bold',
    padding: 'calc(0.4375rem + 1px) 0.4375rem 0.4375rem 0.4375rem',
    lineHeight: '1rem',
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    outline: 0,
    border: `1px solid ${theme.pallete.gray.c60}`,
    borderBottom: `none`,
    background: theme.pallete.gray.c90,
    transition: 'background .15s, border-color .15s',

    '&:not(.disabled):hover': {
      cursor: 'pointer',
      background: theme.pallete.gray.c80,
    },

    '&:not(.disabled):active': {
      background: theme.pallete.gray.c80,
      boxShadow: theme.shadows.inner['10'],
    },

    '&:focus': {
      boxShadow: `inset 0 0 0 2px ${theme.pallete.primary.main}`,
      borderRadius: 4,
    },
  } as CSSProperties,
  active: {
    color: theme.pallete.primary.main,
    background: theme.pallete.gray.c100,
    borderTop: `4px solid ${theme.pallete.primary.main}`,
    paddingTop: 'calc(0.4375rem - 2px)',
    position: 'relative',

    '&::before': {
      content: '""',
      position: 'absolute',
      bottom: '-1px',
      left: '0',
      right: '0',
      height: '1px',
      background: theme.pallete.primary.c100,
    },

    '&:not(.disabled):hover': {
      background: theme.pallete.gray.c90,
    },

    '&:not(.disabled):active': {
      background: theme.pallete.gray.c90,
    },
  } as CSSProperties,
  disabled: {
    color: theme.pallete.text.disabled,

    '&:hover': {
      cursor: 'not-allowed',
    },
  } as CSSProperties,
})
