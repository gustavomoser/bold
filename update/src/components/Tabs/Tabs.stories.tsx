import React from 'react'

import { TabItem } from './TabItem'
import { Tabs } from './Tabs'

export default { title: 'Components/Tabs' }

export const Default = () => (
  <Tabs>
    <TabItem>First item</TabItem>
    <TabItem active>Second item</TabItem>
    <TabItem>Third item</TabItem>
    <TabItem>Fourth item</TabItem>
    <TabItem disabled>Disabled item</TabItem>
  </Tabs>
)
