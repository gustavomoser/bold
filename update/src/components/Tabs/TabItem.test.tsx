import { render } from '@testing-library/react'
import React from 'react'
import { AnchorHTMLAttributes } from 'react'

import { TabItem, TabItemProps } from './TabItem'
import { Tabs } from './Tabs'

it('should render correctly', () => {
  const { container } = render(<TabItem>Item</TabItem>)
  expect(container).toMatchSnapshot()
})

it('should accept "style" prop', () => {
  const { container } = render(<TabItem style={{ color: 'red' }}>Item</TabItem>)
  expect(container).toMatchSnapshot()
})

it('should accept "component" prop', () => {
  const TabLink = (props: AnchorHTMLAttributes<HTMLAnchorElement> & TabItemProps) => (
    <TabItem component='a' {...props} />
  )

  const { container } = render(
    <Tabs>
      <TabLink href='/'>Home</TabLink>
      <TabLink href='/'>Test</TabLink>
      <TabLink href='/' active>
        Active
      </TabLink>
      <TabLink href='/' disabled>
        Disabled
      </TabLink>
    </Tabs>
  )
  expect(container).toMatchSnapshot()
})
