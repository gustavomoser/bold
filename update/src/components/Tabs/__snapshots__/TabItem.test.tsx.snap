// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept "component" prop 1`] = `
.emotion-1 {
  display: inline-block;
}

.emotion-1:not(:first-child) {
  margin-left: 0.25rem;
}

.emotion-0 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #24252E;
  font-weight: bold;
  padding: calc(0.4375rem + 1px) 0.4375rem 0.4375rem 0.4375rem;
  line-height: 1rem;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  outline: 0;
  border: 1px solid #8F8FA2;
  border-bottom: none;
  background: #F0F0F5;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
}

.emotion-0:not(.disabled):hover {
  cursor: pointer;
  background: #D3D4DD;
}

.emotion-0:not(.disabled):active {
  background: #D3D4DD;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-0:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  list-style: none;
  padding: 0;
  margin: 0;
  font-size: 0.875rem;
  border-bottom: 1px solid #D3D4DD;
}

.emotion-4 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #24252E;
  font-weight: bold;
  padding: calc(0.4375rem + 1px) 0.4375rem 0.4375rem 0.4375rem;
  line-height: 1rem;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  outline: 0;
  border: 1px solid #8F8FA2;
  border-bottom: none;
  background: #F0F0F5;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
  color: #0069D0;
  background: #FFFFFF;
  border-top: 4px solid #0069D0;
  padding-top: calc(0.4375rem - 2px);
  position: relative;
}

.emotion-4:not(.disabled):hover {
  cursor: pointer;
  background: #D3D4DD;
}

.emotion-4:not(.disabled):active {
  background: #D3D4DD;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-4:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

.emotion-4::before {
  content: "";
  position: absolute;
  bottom: -1px;
  left: 0;
  right: 0;
  height: 1px;
  background: #FFFFFF;
}

.emotion-4:not(.disabled):hover {
  background: #F0F0F5;
}

.emotion-4:not(.disabled):active {
  background: #F0F0F5;
}

.emotion-6 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #24252E;
  font-weight: bold;
  padding: calc(0.4375rem + 1px) 0.4375rem 0.4375rem 0.4375rem;
  line-height: 1rem;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  outline: 0;
  border: 1px solid #8F8FA2;
  border-bottom: none;
  background: #F0F0F5;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
  color: #AAAAB9;
}

.emotion-6:not(.disabled):hover {
  cursor: pointer;
  background: #D3D4DD;
}

.emotion-6:not(.disabled):active {
  background: #D3D4DD;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-6:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

.emotion-6:hover {
  cursor: not-allowed;
}

<div>
  <ul
    class="emotion-8"
    role="tablist"
  >
    <li
      class="emotion-1"
    >
      <a
        class="emotion-0"
        href="/"
        role="tab"
        tabindex="0"
      >
        Home
      </a>
    </li>
    <li
      class="emotion-1"
    >
      <a
        class="emotion-0"
        href="/"
        role="tab"
        tabindex="-1"
      >
        Test
      </a>
    </li>
    <li
      class="emotion-1"
    >
      <a
        class="emotion-4"
        href="/"
        role="tab"
        tabindex="-1"
      >
        Active
      </a>
    </li>
    <li
      class="emotion-1"
    >
      <a
        class="emotion-6 disabled"
        href="/"
        role="tab"
      >
        Disabled
      </a>
    </li>
  </ul>
</div>
`;

exports[`should accept "style" prop 1`] = `
.emotion-0 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #24252E;
  font-weight: bold;
  padding: calc(0.4375rem + 1px) 0.4375rem 0.4375rem 0.4375rem;
  line-height: 1rem;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  outline: 0;
  border: 1px solid #8F8FA2;
  border-bottom: none;
  background: #F0F0F5;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
}

.emotion-0:not(.disabled):hover {
  cursor: pointer;
  background: #D3D4DD;
}

.emotion-0:not(.disabled):active {
  background: #D3D4DD;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-0:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

.emotion-1 {
  display: inline-block;
  color: red;
}

.emotion-1:not(:first-child) {
  margin-left: 0.25rem;
}

<div>
  <li
    class="emotion-1"
  >
    <span
      class="emotion-0"
      role="tab"
    >
      Item
    </span>
  </li>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-1 {
  display: inline-block;
}

.emotion-1:not(:first-child) {
  margin-left: 0.25rem;
}

.emotion-0 {
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #24252E;
  font-weight: bold;
  padding: calc(0.4375rem + 1px) 0.4375rem 0.4375rem 0.4375rem;
  line-height: 1rem;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  outline: 0;
  border: 1px solid #8F8FA2;
  border-bottom: none;
  background: #F0F0F5;
  -webkit-transition: background .15s,border-color .15s;
  transition: background .15s,border-color .15s;
}

.emotion-0:not(.disabled):hover {
  cursor: pointer;
  background: #D3D4DD;
}

.emotion-0:not(.disabled):active {
  background: #D3D4DD;
  box-shadow: inset 0 1px 4px 0 rgba(0,0,0,0.14),inset 0 2px 4px 0 rgba(0,0,0,0.08);
}

.emotion-0:focus {
  box-shadow: inset 0 0 0 2px #0069D0;
  border-radius: 4px;
}

<div>
  <li
    class="emotion-1"
  >
    <span
      class="emotion-0"
      role="tab"
    >
      Item
    </span>
  </li>
</div>
`;
